### ISSUE TITLE
Cmd: Rotate pic

### ISSUE THREAD

**Issue by bill-mybiz** [ibgib/ibgib#182]
I was thinking i can just include it in metadata to rotate, but that wouldnt change the download file and view file. Download i could conceivably rotate before dl, but view is just serving static file. Hrmm...

**Comment by OvermindDL1** [comment]
You cannot rotate the view of it via a css translate?

**Comment by bill-mybiz** [comment]
Yes, i can rotate the thumbnail in the svg graph no problem. But when you click the "view" pic cmd, it simply points to the file being served (in a new tab/window). I think i could do it via a more customized pic viewer though, but then it no longer would be a simple file being served.  But it might be worth it, since many of the pics i have been uploading are around 8 mb. Otherwise, every rotate would create a duplicate 8 ish mb file.

Thinking beyond rotating tho, what about cropping? Pretty sure that will require a new file, although it is conceivable that this could be done dynamically as well.

Actually, what about any editing with layers? Im not saying i have the skills to build a major editing software, but ibgib is perfectly poised to work with layers and transforms. That might actually make this an interesting design.

Basically right now im hardcoding rendering strategies for the hardcoded "types": text (comments), links (links), and images (pics). But the goal is to have a more dynamic/customizable rendering strategy. I could have a rendering strategy for "complex images" (or whatever) that applies the transforms in the pic^gib data or rel8ns. I think i could dynamically generate client side a new image and set the svg's url to a `data:` url?

**Comment by OvermindDL1** [comment]
You can show a 'part' of an image (the spritesheet method is pretty common in CSS, one image, tons of pictures on it), but that still involves sending the entire image to the user, but once there if it is not the raw file downloaded then it is pretty easy to translate it how you want.

**Comment by bill-mybiz** [comment]
Ah yes, spritesheets! My first foray into c# was with the first xna...i ended up creating my own 3d engine but did quite a few 2d spritesheet examples along the way. That is a **great** visual example of how transforms work in ibGib.

Theyre useful because you can load the expensive asset once in memory and then work with extremely cheap projections (transforms) of the sheet to get the image that you want.

Its also a super example of how ibGib is slightly different and a mindset tweak is required. IbGib is about the _process_ of evolving what would inevitably _become_ a spritesheet. 

So once you upload a pic, that exists "forever". The ibgib is created which contains its own hash, and it points to a file id for the pic that is itself a hash of the picture. These are supposed to be immutable things, otherwise the hash would change.

For example, take my avocado timelapse that im doing, trying to sprout an avocado seed. [Here](https://www.ibgib.com/ibgib/pic%5E2698286D9CD269AF9E5596331899B080A9B8DC24F9C43A2A95A8363DA38836D4) is an ibgib link. If you look at the info for the pic ibgib, it has the following `data`:

```
 "data": {
    "when": "2017-02-14 20:54:31.186739Z",
    "thumb_size": "300x300",
    "thumb_filename": "thumb_88ACA612472CC0CFD0BE5FAD6414E906C395EA2E9973A4DADD84990468AAF97D",
    "thumb_bin_id": "88ACA612472CC0CFD0BE5FAD6414E906C395EA2E9973A4DADD84990468AAF97D",
    "filename": "14871056540322091390889.jpg",
    "ext": ".jpg",
    "content_type": "image/jpeg",
    "bin_id": "50897083C1F28DAF0FCCF202E7400D795EEFB71BCBD80799463D55C7FC089572"
  }
```

The binary id and thumb_bin_id are hashes for the corresponding image files, making it content addressable and verifiable (assuming no sha256 funny business, but we know that will eventually be broken :laughing:)

Binary id: 50897083C1F28DAF0FCCF202E7400D795EEFB71BCBD80799463D55C7FC089572
Url: https://www.ibgib.com/files/50897083C1F28DAF0FCCF202E7400D795EEFB71BCBD80799463D55C7FC089572.jpg

The ibGib is itself hashed and stored in the `gib` field: 2698286D9CD269AF9E5596331899B080A9B8DC24F9C43A2A95A8363DA38836D4 which combined with the `ib` field creates thr ibGib's content addressable URL: https://www.ibgib.com/ibgib/pic%5E2698286D9CD269AF9E5596331899B080A9B8DC24F9C43A2A95A8363DA38836D4 .

So one of my early design decisions was that "content" (e.g. binaries) is expensive and "pointers" (e.g. ib^gib addresses) are cheap. So here is the trippy part that you just helped me understand: _All images act togeher as a single big spritesheet._ Obviously the performance is not the same because were thinking of a larger, distributed computer, but the similarity is that they all live in a sparsely populated sha256(ish) sized space, and any ibGib can act as the concrete projection that translates to a single rendered image.

Yes thats very interesting indeed! 😃

I'm going to let that digest and hammock drive some development.

**Comment by OvermindDL1** [comment]
Hah, that would indeed fit very well!

But yeah, a lot of sites use spritesheets for their images, with the CSS showing only part of an image and they just assign the CSS where-ever they need.  Downloads faster, uses less memory, etc...  :-)

**Comment by bill-mybiz** [comment]
Well, I've decided to go ahead with the more interesting approach of actually editing the pic instead of only the rotate pic aspect. Rotate will still be the first thing to implement, but it will be done via a transform process/custom rendering of an ibGib instead of "physically" rotating the picture content and saving a new file. 

So I've removed the "light" label and edited the name of the issue. 

That said, once I get a couple things of a higher priority implemented (currently working on notifications and related), when I revisit this issue I'm going to sketch out the new rendering process. Thanks for your input @OvermindDL1 :clap: and I'm sure you'll see the notifications when I've posted a big thing on the algorithm for actually implementing this.
